### PR TITLE
Better WICD dependency handling

### PR DESCRIPTION
**Pull request by sebsoto**
Includes changes to both the unit test code and functionality, improving the way WICD handle service dependencies

### CODE REVIEW

File: pkg/daemon/controller/controller.go [hunk 1]
```diff
@@ -189,8 +189,12 @@ func (sc *ServiceController) reconcileService(service winsvc.Service, expected s
 	}
 
 	if !reflect.DeepEqual(config.Dependencies, expected.Dependencies) {
-		config.Dependencies = expected.Dependencies
-		updateRequired = true
+		// reflect.DeepEqual considers a nil slice not equal to an empty slice, so we need to include an extra check
+		// to ensure they both aren't of length 0, which is desired.
+		if !(len(config.Dependencies) == len(expected.Dependencies) && len(config.Dependencies) == 0) {
```
Comment (saifshaikh48) [inline pkg/daemon/controller/controller.go:194]: Also, consider adding a unit test around this.

Comment (selansen) [reply pkg/daemon/controller/controller.go:194]: +1

Comment (sebsoto) [reply pkg/daemon/controller/controller.go:194]: done

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

File: pkg/daemon/controller/controller.go [hunk 2]
```diff
@@ -308,3 +308,16 @@ func currentNode(nodes *core.NodeList, localAddrs []net.Addr) (*core.Node, error
...
 }
+
+// slicesEquivalent returns true if the slices have the same content, or if they both have no content
+func slicesEquivalent(s1, s2 []string) bool {
+	if reflect.DeepEqual(s1, s2) {
+		return true
+	}
+	// reflect.DeepEqual considers a nil slice not equal to an empty slice, so we need to include an extra check
+	// to see if they both length zero, one being nil, and one being empty
+	if len(s1) == len(s2) && len(s1) == 0 {
```
Comment (jrvaldes) [inline pkg/daemon/controller/controller.go:319]: No need for `if` statement, you can `return len(s1) == len(s2) && len(s1) == 0`

File: pkg/daemon/winsvc/winsvc.go [hunk 3]
```diff
@@ -173,6 +184,33 @@ func (f *FakeService) Control(cmd svc.Cmd) (svc.Status, error) {
...
+		// scenario.
+		existingServices := f.serviceList.listServiceNames()
+		for _, serviceName := range existingServices {
+			service, present := f.serviceList.read(serviceName)
+			if !present {
+				return svc.Status{}, errors.New("unable to open service " + serviceName)
+			}
+			config, err := service.Config()
+			if err != nil {
+				return svc.Status{}, errors.Wrap(err, "error getting service config")
```
Comment (jrvaldes) [inline pkg/daemon/winsvc/winsvc.go:198]: Would be nice to include the `serviceName` in the err message.

**Comment by jrvaldes** [comment]
LGTM, @sebsoto  feel free to remove the hold once you review the comments. 

/hold

File: pkg/daemon/controller/controller.go [hunk 4]
```diff
@@ -188,7 +188,7 @@ func (sc *ServiceController) reconcileService(service winsvc.Service, expected s
 		updateRequired = true
 	}
 
-	if !reflect.DeepEqual(config.Dependencies, expected.Dependencies) {
+	if !slicesEquivalent(config.Dependencies, expected.Dependencies) {
 		config.Dependencies = expected.Dependencies
 		updateRequired = true
 	}
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply pkg/daemon/controller/controller.go:194]: There are multiple places where updateRequired can be set, so will leave as is

**Pull request merged**